redis/src/cluster_async/connections_logic.rs

```diff
@@ -53,48 +54,406 @@ pub(crate) async fn get_or_create_conn<C>(
+            ConnectAndCheckResult::Success(_) => None,
+            ConnectAndCheckResult::ManagementConnectionFailed { err, .. } => Some(err),
+            ConnectAndCheckResult::Failed(err) => Some(err),
+        }
+    }
+}
+
+impl<C> From<RedisError> for ConnectAndCheckResult<C> {
+    fn from(value: RedisError) -> Self {
+        ConnectAndCheckResult::Failed(value)
+    }
+}
+
+impl<C> From<AsyncClusterNode<C>> for ConnectAndCheckResult<C> {
+    fn from(value: AsyncClusterNode<C>) -> Self {
+        ConnectAndCheckResult::Success(value)
+    }
+}
+
+impl<C> From<RedisResult<AsyncClusterNode<C>>> for ConnectAndCheckResult<C> {
+    fn from(value: RedisResult<AsyncClusterNode<C>>) -> Self {
+        match value {
+            Ok(value) => value.into(),
+            Err(err) => err.into(),
+        }
     }
 }
 
 #[doc(hidden)]
 pub async fn connect_and_check<C>(
+    addr: &str,
+    params: ClusterParams,
+    socket_addr: Option<SocketAddr>,
+    conn_type: RefreshConnectionType,
+    node: Option<AsyncClusterNode<C>>,
+) -> ConnectAndCheckResult<C>
+where
+    C: ConnectionLike + Connect + Send + Sync + 'static + Clone,
+{
+    match conn_type {
+        RefreshConnectionType::OnlyUserConnection => {
+            let (user_conn, ip) =
+                match create_user_connection(addr, params.clone(), socket_addr).await {
+                    Ok(tuple) => tuple,
+                    Err(err) => return err.into(),
+                };
+            if let Some(node) = node {
+                let mut management_conn = match node.management_connection {
+                    Some(ref conn) => Some(conn.clone().await),
+                    None => None,
+                };
+                if ip != node.ip {
+                    // New IP was found, refresh the management connection too
+                    management_conn =
+                        create_and_setup_management_connection(addr, params, socket_addr)
+                            .await
+                            .ok()
+                            .map(|(conn, _ip): (C, Option<IpAddr>)| conn);
+                }
+                create_async_node(user_conn, management_conn, ip).into()
+            } else {
+                create_async_node(user_conn, None, ip).into()
+            }
+        }
+        RefreshConnectionType::OnlyManagementConnection => {
+            // Refreshing only the management connection requires the node to exist alongside a user connection. Otherwise, refresh all connections.
+            match node {
+                Some(node) => {
+                    connect_and_check_only_management_conn(addr, params, socket_addr, node).await
+                }
+                None => connect_and_check_all_connections(addr, params, socket_addr).await,
+            }
+        }
+        RefreshConnectionType::AllConnections => {
+            connect_and_check_all_connections(addr, params, socket_addr).await
+        }
+    }
+}
+
+async fn create_user_connection<C>(
```

Comment:
create_and_setup_management_connection and create_user_connection names should be the same in all but the connection name (user/management)